### PR TITLE
Work Submission

### PR DESCRIPTION
**Pull request by AnanthVivekanand**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -12,9 +13,14 @@ const defaultConfig = {
 };
 
 class Client {
-
+submit() {
+var args = Array.prototype.slice.call(arguments);
```
Comment (arnabk) [inline src/index.js]: I think we just do `submit(...args)` instead of using slice.
I am not sure why the code formatting is messed up all over. Could you please check?

**Comment by AnanthVivekanand** [comment]
Ah, I was trying to prepend `client` (the socket) to the arguments. That way, when the arguments are forwarded to submitWork.js, it will have access to the socket in order to send data to the pool. Without adding `client` to the arguments, it won't be able to access the socket.

It works fine, do you prefer a different method? 

```
submit() {
    var args = Array.prototype.slice.call(arguments);
    args.unshift(this.client); //Adds client to the arguments
    submitWork.apply(this, args); //Forwards the arguments (and client) to submitWork.js
}
```

**Comment by arnabk** [comment]
@KorkyMonster There are couple of things to consider.

- It looks like you are using an IDE which uses 4 spaces as tab. This project has `.editorconfig` which defines what this project is formatted with. It would be better if your IDE uses that instead of overriding it.
- I took the time to look into your code further. I think it would be better if you define what arguments you are passing instead of `...args`. Using `arguments` is powerful but not very useful to keep the code readability. Also, instead of passing each argument as separate parameter, pass them in one object like opts. That way, the order of the argument wont matter. Generally what I do is, when the list of parameter is 2 or less, I pass them as separate parameter, but when it grows, its better to use `opt` for flexibility later.

File: src/processData.js [hunk 2]
```diff
@@ -2,75 +2,86 @@ const extend = require('lodash/extend');
...
+    onAuthorizeFail,
+    onSubscribe,
+    onNewDifficulty,
+    worker,
+    password,
+    onNewMiningWork,
+    onSubmitWorkSuccess,
+    onSubmitWorkFail,
+  } = updatedOptions;
+  console.log(jsonData);
```
Comment (arnabk) [inline src/processData.js]: Please remove this `console.log` statement

File: src/submitWork.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+const {
+  submitWork
+} = require('./messageContants');
+
+module.exports = (options, client) => {
```
Comment (arnabk) [inline src/submitWork.js]: Once you do the above changes, client will be part of options.

File: src/index.js [hunk 4]
```diff
@@ -12,9 +13,12 @@ const defaultConfig = {
 };
 
 class Client {
-
+  submit(options) {
```
Comment (arnabk) [inline src/index.js:19]: Can we  change this code to use 
```
submit(options) {
  submitWork({
     ...options,
     client, 
  });
}
```

**Comment by arnabk** [comment]
If you are interested, you could use destructing in lot of places and improve the code. If not, I will do it in separate PR.

File: src/index.js [hunk 5]
```diff
@@ -14,7 +14,10 @@ const defaultConfig = {
 
 class Client {
   submit(options) {
-    submitWork.apply(this, [options, client]);
+    submitWork([
+      ...options,
+      this.client,
```
Comment (arnabk) [inline src/index.js]: This won't work. You should add following code in the top of the module

```
const client = new net.Socket();
client.setEncoding('utf8');
```
instead of `start` method

File: src/submitWork.js [hunk 6]
```diff
@@ -2,6 +2,6 @@ const {
   submitWork
 } = require('./messageContants');
 
-module.exports = (options, client) => {
-  client.write(submitWork.replace("<worker.name>", options.worker_name).replace("<jobID>", options.job_id).replace("<ExtraNonce2>", options.extranonce2).replace("<ntime>", options.ntime).replace("<nonce>", options.nonce));
+module.exports = (options) => {
+  options[5].write(submitWork.replace("<worker.name>", options[0]).replace("<jobID>", options[1]).replace("<ExtraNonce2>", options[2]).replace("<ntime>", options[3]).replace("<nonce>", options[4]));
```
Comment (arnabk) [inline src/submitWork.js]: This is bad. Use object destructing.

**Comment by AnanthVivekanand** [comment]
Is this project still being maintained? PR has been made a while ago to solve issue #2.

File: src/connect.js [hunk 7]
```diff
@@ -1,10 +1,16 @@
-const { subscribe } = require('./messageContants');
+const {
```
Comment (arnabk) [inline src/connect.js]: This is an unnecessary change. Please revert changes to this file

File: src/onError.js [hunk 8]
```diff
@@ -1,12 +1,15 @@
 const connect = require('./connect');
 
 module.exports = (client, options, error) => {
-  const { autoReconnectOnError, onError } = options;
+  const {
+    autoReconnectOnError,
```
Comment (arnabk) [inline src/onError.js]: Please revert changes to this file.

**Comment by arnabk** [comment]
It is. You have to follow the coding standards of the project and do not submit unnecessary formatting. Please go over list of comments and do the changes accordingly.

**Comment by AnanthVivekanand** [comment]
I've cleaned up several cases of unnecessary formatting. Thanks for letting me know about that.

I've left `src/processData.js` with some formatting changes because `.editorconfig` declares this:
```
indent_style = space
indent_size = 2
```

I've changed the instances of 4 spaces per indent in that files to 2, making it adhere to .editorconfig. Please look over the changes I've made and let me know of what you think.

File: src/processData.js [hunk 9]
```diff
@@ -2,75 +2,86 @@ const extend = require('lodash/extend');
...
-            onSubscribe({
-                extraNonce1: workObject.extraNonce1,
-                extraNonce2Size: workObject.extraNonce2Size,
-            });
-        }
-        if (worker) {
-            client.write(authorize.replace("<worker.name>", worker).replace("<worker.pass>", defaultTo(password, 'x')));
-        }
-        break;
-      case miningNotify:
```
Comment (AnanthVivekanand) [inline src/processData.js:56]: I've changed instances of 4 spaces per indent, like that below, to two space per indent.

File: src/connect.js (hunk 7 above)
Comment (AnanthVivekanand) [reply src/connect.js]: Changes have been reverted.

File: src/onError.js (hunk 8 above)
Comment (AnanthVivekanand) [reply src/onError.js]: Changes have been reverted.

**Comment by arnabk** [comment]
@KorkyMonster Looks good to me. Have you tested connecting to a pool and making sure that it is receiving work and is also able to submit work? If yes, then please run `npm version minor` and push the changes. I will merge and release it.

Thank you

**Comment by AnanthVivekanand** [comment]
Published! 

Here's a javascript miner written with `stratum-client`: https://github.com/KorkyMonster/garlicoin-miner-nodejs/

**Comment by arnabk** [comment]
Fixes https://github.com/arnabk/stratum-client/issues/2
